tests/serial/time/implicit/parallel-coupling/WaveformSubcyclingWithDifferentNumberOfDts.cpp

```diff
@@ -0,0 +1,119 @@
+ */
+BOOST_AUTO_TEST_CASE(WaveformSubcyclingWithDifferentNumberOfDts)
+{
+  PRECICE_TEST("SolverOne"_on(1_rank), "SolverTwo"_on(1_rank));
+
+  Participant precice(context.name, context.config(), 0, 1);
+
+  typedef double (*DataFunction)(double);
+
+  // The functions that are exchanged in iteration 1 and 2 respectively.
+  DataFunction funcIterOne = [](double t) -> double {
+    return (double) (1 + 2 * t) * (2 + 3 * t) * (3 + 4 * t);
+  };
+  DataFunction funcIterTwo = [](double t) -> double {
+    return (double) (6 + 5 * t) * (1 + 6 * t) * (1 + 6 * t);
+  };
+
+  BOOST_TEST(funcIterOne(0) == funcIterTwo(0)); // because the data at WINDOW_START does not change over iterations.
+
+  int         nTimeSteps;
+  std::string meshName, writeDataName, readDataName;
+  if (context.isNamed("SolverOne")) {
+    meshName      = "MeshOne";
+    writeDataName = "DataOne";
+    readDataName  = "DataTwo";
+    nTimeSteps    = 6;
+  } else {
+    BOOST_TEST(context.isNamed("SolverTwo"));
+    meshName      = "MeshTwo";
+    writeDataName = "DataTwo";
+    readDataName  = "DataOne";
+    nTimeSteps    = 4;
+  }
+
+  double   writeData, readData;
+  VertexID vertexID;
+
+  double v0[] = {0, 0, 0};
+  vertexID    = precice.setMeshVertex(meshName, v0);
+  int    timestepCheckpoint;
+  double time = 0;
+
+  if (precice.requiresInitialData()) {
+    writeData = funcIterOne(time);
+    precice.writeData(meshName, writeDataName, {&vertexID, 1}, {&writeData, 1});
+  }
+
+  precice.initialize();
+  double TimeWindowSize = precice.getMaxTimeStepSize();
+  double dt             = TimeWindowSize / nTimeSteps;
+  double timeCheckpoint;
+  int    iterations;
+
+  while (precice.isCouplingOngoing()) {
+    if (precice.requiresWritingCheckpoint()) {
+      dt             = TimeWindowSize / nTimeSteps;
+      timeCheckpoint = time;
+      iterations     = 0;
+    }
+    precice.readData(meshName, readDataName, {&vertexID, 1}, dt, {&readData, 1});
+
+    if (iterations == 0) { // in the first iteration of each window, we only have one sample of data. Therefore constant interpolation
+      BOOST_TEST(readData == funcIterOne(timeCheckpoint));
+    } else if (iterations == 1) { // Check that we receive the first function in the second iteration
+      BOOST_TEST(readData == funcIterOne(time + dt));
+    } else { // Check that we receive the second function in the third iteration
+      BOOST_TEST(readData == funcIterTwo(time + dt));
+    }
+
+    // solve usually goes here. Dummy solve: Just sampling the two functions.
+    time += dt;
+    if (iterations == 0) { // Sends the first function in the first iteration
+      writeData = funcIterOne(time);
+    } else { //Otherwise send the second function
+      writeData = funcIterTwo(time);
+    }
+
+    precice.writeData(meshName, writeDataName, {&vertexID, 1}, {&writeData, 1});
+    precice.advance(dt);
+    double maxDt = precice.getMaxTimeStepSize();
+    dt           = dt > maxDt ? maxDt : dt;
```

Comment:
Please name consistently:

https://github.com/precice/precice/blob/f1adf55c6a34bb7b6e87e7691f69785cf0558741/tests/serial/time/implicit/serial-coupling/ReadWriteScalarDataWithWaveformSamplingFirst.cpp#L103-L104

Also important: You are overwriting the `dt` defined by the solver here. This can lead to hard-to-detect errors. I therefore, recently updated our documentation to make sure to always explicitly differentiate between the dt requested by the solver, the dt requested by preCICE, and the dt that is finally used. See https://github.com/precice/precice.github.io/pull/245/files.